### ISSUE TITLE
Help to locate the performance bottleneck of Pachyderm

### ISSUE THREAD

**Issue by ShengjieLuo** [pachyderm/pachyderm#701]
Refer to [https://github.com/pachyderm/pachyderm/issues/699](url) . It seems that I didn't give a clear expression about performance problem, as well as the speculation about FUSE is not correct. So I do some comparison test today and repost this issue.

I run the word count sample in which data is come from the wiki page just like the official sample mentioned. 

```
"cmd": [ "wget",
        "-e", "robots=off",
        "--recursive",
        "--level", "1",
        "--adjust-extension",
        "--no-check-certificate",
        "--no-directories",
        "--directory-prefix",
        "/pfs/out",
        "https://en.wikipedia.org/wiki/Main_Page"
```

First Equipment: I use AWS m4.Xlarge vm + ubuntu14.04 + Kubernetes 1.2.2 + Pachyderm 1.1.0 to run the wourd count example. Note that there is only one node. The whole process uses 29 minutes and the reduce process uses 24 minutes.

```
ID                                 OUTPUT                                              STARTED             DURATION             STATE
b80331f28ac956ba308011b7ef969054   wordcount_reduce/605aa77a63f4456f867f2c476aced43d   59 minutes ago      24 minutes           success
105a07fead7878072e178a9440e5134f   wordcount_map/366ed6c782024fe0a58e5b43346ea70f      About an hour ago   5 minutes            success
cdeca10ac3f24292824a5c878542517b   wordcount_input/30e1f6d2f41a42eca350f76f15722d8b    About an hour ago   18 seconds           success
```

Second Equipment: I use AWS m4.XLarge + ubuntu14.04 to run the example directly and locally, of which the CPU number and memory size is the same with that of Pachyderm node. The  map.go  program and reduce shell script used in Pachyderm sample are also involved in the Local server processing. Finally, the local server uses about 2.5 minutes to finish the task. The reduce task takes the major part. See my time statistical result here.

```
begin time: Fri Aug  5 07:04:26 2016
cat: ./map: Is a directory
end time: Fri Aug  5 07:06:49 2016
```

Third Equipment: I use AWS m4.XLarge + ubuntu14.04 + Hadoop2.7.2 to run the example. A typical wordcount java program in Hadoop is used to be compared with Pachyderm Pipeline. Unfortunately, Hadoop uses less than 10 seconds although I set the Hadoop as the single-node-non-distribution mode, i.e only one Java process is used.

In conclusion, since the Hadoop Word Count use a different method to realize counting, also , Hadoop is a mature data platform. Please just forget it. Let's focus on the performance difference between local server and Pachyderm Pipeline (2.5minutes vs 30 minutes) to find the performance bottleneck.
Could you give me some idea about the performance problem? I have excluded some impossible reasons.
1. Docker Pull : The docker image is build before experiment. And the docker load is so quick.
2. Sources shortage: Don't be worried about that. The 300 GB disk only spends less than 3%. All of the `tmpfs` has a large space. The memory size is also enough for that.
3. Pipeline Programming : All of the execute command follow the example guide strictly.
4. Service Down : I set a daemon to check whether there is any kubernetes or pachyderm service down during the process. And all services work normally.

Several speculations are given:

1 Filesystem Problem: For a simple computational task, Pachyderm doesn't meet a fatal performance problem.  But in word count example, there are many documents (exactly one word one file), does the frequent file write weaken the performance? Or the docker operation on file is slower than local host?

2  Message Send:  The result of `wordcount_map` is needed to be sent to `wordcount_reduce`. As I observe, this process is not done by memory, I am not sure it is fulfilled by DISK file or RESTful API.  Probably this step would cause a time loss.

**Comment by ShengjieLuo** [comment]
Note: In the second experiment, i.e the local server experiment. I do not use the Pachyderm nor Hadoop. I launch a new server on AWS, download the original data, execute the `map.go` program to do map task and execute the `find /pfs/wordcount_map -name '*' | while read count; do cat $count | awk '{ sum+=$1} END {print sum}' >/tmp/count; mv /tmp/count /pfs/out/`basename $count`; done` to do reduce task. All these steps are manually fulfilled.  And a timer is set to calculate the processing time.

**Comment by zyluo** [comment]
Based on your first and second experiment, the Pachyderm's job shim seems to be the bottleneck.

**Comment by jdoliner** [comment]
Hmm, I'm thinking the most likely culprit here is filesystem latency. If I understand your tests correctly the local version is using a local filesystem while the version in amazon is using s3. That means that the reduce stage needs to do a lot of writes in s3, 1 per word that it counts which would be pretty slow. In the future this will be optimized so that those writes can be packed into a single s3 write which should greatly increase the speed.

**Comment by derekchiang** [comment]
@ShengjieLuo thanks for bringing the performance issues to our attention!  So far we've been spending most our engineering efforts on the correctness of the system , and not as much on performance, so I'm sure there's a lot of room for improvement there.

I agree with @jdoliner that the performance difference between experiment 1 and 2 is mostly due to the frequent writes to S3.  In general, Pachyderm is much better at handling large files than handling many small files, as is the case for the wordcount example, where each unique word results in a small file.

As for the performance difference between 2 and 3 (local vs Hadoop), I'd attribute it to the fact that both the `map.go` program and the `reduce` bash script make use of the local file system, while the Hadoop program probably does the bulk of the work in memory.  Note that in theory, Pachyderm can run in memory as well: https://github.com/pachyderm/pachyderm/issues/546

**Comment by ShengjieLuo** [comment]
@jdoliner Thanks you for your advice. I use general SSD (gp2) disk for AWS VM, also the dick operation is regarded as the most possible bottleneck in our team. I still have some further problems. As you have noticed, Pachyderm will be optimized so that writing operation can be packed into a single S3, does it mean that currently Pachyderm use more than one S3 volume to work even I only setup one docker container to fulfill the reduce task?

**Comment by ShengjieLuo** [comment]
A table is given to show the performance test more clearly
![performancetest](https://cloud.githubusercontent.com/assets/20234415/17501174/b87d6308-5e0e-11e6-95ab-58f78baa52fc.png)

**Comment by derekchiang** [comment]
Nice table!  But just to be clear, in case someone thinks that Hadoop is 174 times faster than Pachyderm: it's an unfair comparison because the Hadoop setup runs purely in memory, while the Pachyderm setup sends data over the network.  It'd be more fair if the Hadoop setup was also writing data to a remote location.

**Comment by derekchiang** [comment]
@ShengjieLuo oh and for your follow-up question: Pachyderm only uses one S3 bucket by default.  It's just that currently each time we write a small file, we are sending a write request over the network.  What @jdoliner was suggesting was that we can pack the write requests into one big write request, so as to reduce latency introduced by the network.

**Comment by ShengjieLuo** [comment]
@derekchiang For a single-node Hadoop, the operation is based on memory and local file system. However, let's focus on the difference of local server and Pachyderm first. It's important to give a detailed clarification here.

The aws has an automatic method to deploy Kubernetes cluster, and Pachyderm can be deployed on this five-node kubernetes cluster. In this case, S3 bucket is used to transform the data from one node to the other node.

However in my experiment, I launched an instance on AWS. and launch the kubernetes single-node cluster on this instance. The AWS server plays a role the same with a local server.
The script used is 
```
docker run \
    -d \
    --volume=/:/rootfs:ro \
    --volume=/sys:/sys:ro \
    --volume=/dev:/dev \
    --volume=/var/lib/docker/:/var/lib/docker:rw \
    --volume=/var/lib/kubelet/:/var/lib/kubelet:rw \
    --volume=/var/run:/var/run:rw \
    --net=host \
    --pid=host \
    --privileged=true \
    gcr.io/google_containers/hyperkube:v1.2.2 \
    /hyperkube kubelet \
        --containerized \
        --hostname-override="127.0.0.1" \
        --address="0.0.0.0" \
        --api-servers=http://localhost:8080 \
        --config=/etc/kubernetes/manifests \
        --allow-privileged=true
until kubectl version 2>/dev/null >/dev/null; do sleep 5; done  
```
See kubernetes node here,
![image](https://cloud.githubusercontent.com/assets/20234415/17501991/c2cb49da-5e16-11e6-9f1b-f072c1959a24.png)
See kubernetes work status here, 
![image](https://cloud.githubusercontent.com/assets/20234415/17501997/d3a6fda8-5e16-11e6-8d43-e69b512ed027.png)

After kubernetes has been launched, I deploy Pachyderm by,
```
cd $GOPATH/src/github.com/pachyderm/pachyderm
make launch
```
In conclusion, the aws vm used in this experiment is just like a local server. Does S3 problem still exist in this step?

**Comment by derekchiang** [comment]
@ShengjieLuo ah, I think I misunderstood what you were saying before.  Somehow I thought that the first experiment involved writing to S3, but from what you just said, that doesn't seem to be the case.

So just to be clear:

* In the first experiment, you were running Pachyderm on a single node, with no S3 involved.  This means that Pachyderm was writing data to the local disk.
* In the second experiment, you were running the `map` and `reduce` programs manually, on a single node.

If that's the case, then what you found out is that adding the Pachyderm layer alone causes the processing time to jump from 2.5 min to 29 min, is that correct?  If that's the case, that's certainly very concerning, and we should be looking into this as soon as possible.

**Comment by ShengjieLuo** [comment]
@derekchiang I would repeat the test as soon as possible to verify the test result. Thanks for your help

**Comment by derekchiang** [comment]
@ShengjieLuo thanks for taking the time to evaluate Pachyderm!

**Comment by ShengjieLuo** [comment]
@derekchiang The experiment has been repeated today, no significant difference between the experiment yesterday.
![image](https://cloud.githubusercontent.com/assets/20234415/17509706/f0f2129e-5e4d-11e6-92e8-1c35c5cea793.png)
